static/js/color-light.js

```diff
@@ -0,0 +1,265 @@
+/**
+ * Color Bulb.
+ *
+ * UI element representing a bulb with control over its color
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+'use strict';
+
+/* globals OnOffSwitch, Thing, ThingDetailLayout */
+
+/**
+ * ColorLight Constructor (extends OnOffSwitch).
+ *
+ * @param Object description Thing description object.
+ * @param {String} format 'svg', 'html', or 'htmlDetail'.
+ */
+function ColorLight(description, format) {
+  Thing.call(this, description, format);
+  if (format == 'svg') {
+    // For now the SVG view is just a link.
+    return this;
+  }
+  this.onPropertyUrl = new URL(this.propertyDescriptions.on.href, this.href);
+  this.colorPropertyUrl = new URL(this.propertyDescriptions.color.href,
+                                  this.href);
+
+  this.updateStatus();
+  this.colorLight = this.element.querySelector('.color-light');
+  this.colorLightIconPath =
+    this.element.querySelector('.color-light-icon-path');
+
+  if (format === 'htmlDetail') {
+    this.onOffSwitch = this.element.querySelector('.color-light-onoff-switch');
+    this.onOffLabel = this.element.querySelector('.color-light-onoff-label');
+    this.colorInput = this.element.querySelector('.color-light-color');
+
+    this.onOffSwitch.addEventListener('click', this.handleClick.bind(this));
+    this.colorInput.addEventListener('change', () => {
+      this.setColor(this.colorInput.value);
+    });
+
+    this.layout = new ThingDetailLayout(
+      this.element.querySelectorAll('.thing-detail-container'));
+  } else {
+    this.element.addEventListener('click', this.handleClick.bind(this));
+  }
+  return this;
+}
+
+ColorLight.prototype = Object.create(OnOffSwitch.prototype);
+
+ColorLight.prototype.iconView = function() {
+  let color = this.properties.color;
+  return `<div class="color-light" style="background: ${color};">
```

Comment:
I feel like there's a better way of doing this rather than hard coding the SVG markup in JavaScript. Could the icon be a background image or an embedded SVG file for example (it's a bit tricky to tell without being able to inspect the live HTML)?

Reply:
IIRC the icon has to be in the HTML for its stroke and fill to be modified by classes/js. The main things that I'm pretty sure will not work are making it a background image or using `<img src="path-to-svg.js">`.